### PR TITLE
Dont generate preview if no uploads

### LINKED ISSUE #254: What does "generate new preview version" do?
In the old GIDS uploaded files were not merged into the bigger institution until a user told the DS to merge all the files together. The purpose, as I understood it, was so that if a file was uploaded that had errors there would be a chance to do it again before merging it all together.

However, now the "generate new preview version" is now in a different place and has a user warning that says, "Are you sure?"

image

image

At what point are CSVs merged into the larger institution file? And if it is when you select this "generate new preview version" button shouldn't it be up with the other three action buttons?

image

@mphprogrammer @saneshark

### PR DESCRIPTION
**Pull request by saneshark**
- optimized query for latest_uploads
- added helper method on version to determine if buildable?
- updated views to Fix #254
- added a check to build action to ensure buildable or to raise flash error.

### CODE REVIEW

**Comment by saneshark** [comment]
This probably needs specs!

File: app/models/version.rb [hunk 1]
```diff
@@ -26,6 +26,20 @@ def self.current_preview
...
+  def self.buildable?
+    Version.buildable_state.to_s.match('can_create')
+  end
+
+  def self.buildable_state
+    upload_dates = Upload.last_uploads.to_a.map(&:updated_at)
+    preview = Version.current_preview
+    return :not_enough_uploads if upload_dates.length < InstitutionBuilder::TABLES.length
+    return :too_many_uploads if upload_dates.length > InstitutionBuilder::TABLES.length
+    return :can_create_first_preview if preview.nil?
```
Comment (patrickvinograd) [inline app/models/version.rb]: Good catch!

File: app/controllers/dashboards_controller.rb [hunk 2]
```diff
@@ -54,6 +55,12 @@ def push
 
   private
 
+  def verify_buildable
+    return true if Version.buildable?
+    flash.alert = 'Cannot build a new version since no new uploads have been made.'
```
Comment (patrickvinograd) [inline app/controllers/dashboards_controller.rb]: I would change this to "Cannot build a new version since no new CSVs have been uploaded."

**Review by patrickvinograd — APPROVED** [review]
LGTM, with one optional alert phrasing suggestion.

**Review by mphprogrammer — APPROVED** [review]
LGTM - like the query

**Pull request merged**